### PR TITLE
Update model classes to use arraylias

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Summary

Updates the model classes to use arraylias.

Files to update:
- ~generator_model.py~
   - ~test_generator_model.py~
- ~hamiltonian_model.py~
   - ~test_hamiltonian_model.py~
- ~lindblad_model.py~
   - ~test_lindblad_model.py~

Interface changes:
- `copy` method has been removed
- Setters for everything other than `signals` and `in_frame_basis` are being removed. I had wanted to make these classes fully immutable but the ability to set these two properties is used quite a bit throughout the solver infrastructure and I don't want to additionally make a change of this scale at this time. (In any case the main mutability issue had to do with the operators, so for now it's okay to keep `signals` and `in_frame_basis` as mutable.)
- Changed `evaluation_mode` to `array_library`
    - For `LindbladModel`, will also add a “vectorized” boolean
- To use `array_library=“scipy_sparse”` or `"jax_sparse"`, a user MUST pass this as `array_library`.
   - This is how it used to be before so it isn't such a big deal - the user will need to pass the arrays in dense format, then they will be converted to sparse format internally. This isn't desirable but I'm willing to accept it for now as it would only be an issue for extremely large models.

Other to do:
- Add breaking release note for these changes

### Details and comments

General notes:
- I changed around the way that `HamiltonianModel` relies on the methods `GeneratorModel`, but the actual user-side behaviour has not changed (aside from the same changes made to `GeneratorModel`).

Test files that need to run before merge:
- `test.dynamics.models.test_lindblad_model`
- `test.dynamics.models.test_hamiltonian_model`
- `test.dynamics.models.test_generator_model`
- `test.dynamics.models.rotating_frame`
- `test.dynamics.models.operator_collections`
- `test.dynamics.arraylias` (whole folder)

### CODE REVIEW

File: test/dynamics/models/test_lindblad_model.py
Comment (to24toro) [inline test/dynamics/models/test_lindblad_model.py:1]: It appears that the type of the Lindblad model, as in model(1.0, rho), is not being checked using self.assertArrayType(). is this considered acceptable?

**Review by to24toro — COMMENTED** [review]
Thanks, @DanPuzzuoli .
I added some minor change suggestions.

File: test/dynamics/models/test_hamiltonian_model.py [hunk 1]
```diff
@@ -130,16 +135,16 @@ def _basic_frame_evaluate_test(self, frame_operator, t):
...
 
     def test_diag_frame_operator_basic_hamiltonian(self):
         """Test setting a diagonal frame operator for the internally
         set up basic hamiltonian.
         """
 
-        self._basic_frame_evaluate_test(Array([1.0, -1.0]), 1.123)
-        self._basic_frame_evaluate_test(Array([1.0, -1.0]), np.pi)
+        self._basic_frame_evaluate_test(np.array([1.0, -1.0]), 1.123)
+        self._basic_frame_evaluate_test(np.array([1.0, -1.0]), np.pi)
```
[the remark this answers is not in the document]
Comment (to24toro) [reply test/dynamics/models/test_hamiltonian_model.py:147]: Need to convert `np.array()` to `self.asarray()`?

File: test/dynamics/models/test_lindblad_model.py
Comment (DanPuzzuoli) [reply test/dynamics/models/test_lindblad_model.py:1]: Yes the issue with using `assertArrayType` here is that the output of `model(1.0, rho)` will always be dense, even if the model is internally using sparse arrays (sparse x dense multiplication produces dense).

File: test/dynamics/models/test_hamiltonian_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply test/dynamics/models/test_hamiltonian_model.py:147]: So the `frame_operator` in rotating frame actually needs to be dense. The code within explicitly assumes this, so we can't use `self.asarray`.

File: qiskit_dynamics/models/lindblad_model.py [hunk 2]
```diff
@@ -660,61 +557,39 @@ def construct_lindblad_operator_collection(
...
         static_dissipators: Dissipation operators with coefficient 1.
         dissipator_operators: Dissipation operators with variable coefficients.
 
     Returns:
-        BaseLindbladOperatorCollection: Right-hand side evaluation object.
-
-    Raises:
-        NotImplementedError: If ``evaluation_mode`` is not one of the above supported evaluation
-            modes.
+        OperatorCollection: Right-hand side evaluation object.
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/models/lindblad_model.py:568]: Yeah - here I was lazy about listing them all

[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/models/lindblad_model.py:568]: I've just updated the doc string to show the proper `Union` type here.

**Comment by DanPuzzuoli** [comment]
I've added a release note as well.

**Review by to24toro — APPROVED** [review]
LGTM!

**Pull request merged**